### PR TITLE
feature: Dataset dependencies in separate fields

### PR DESCRIPTION
**Pull request by arturowczarek**
* When flag columnLineage.deprecatedMechanismEnabled, then the dataset dependencies are extracted from property field into separate field (dataset)

### Problem

The dataset dependencies are mixed with field dependencies

Relates to: #3084
This is a continuation of: https://github.com/OpenLineage/OpenLineage/pull/3098
The next change is #3100

### Solution

We should have the dataset dependencies separate

> **Note:** All schema changes require discussion. Please [link the issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) for context.

- [ ] Your change modifies the [core](https://github.com/OpenLineage/OpenLineage/blob/main/spec/OpenLineage.json) OpenLineage model
- [X] Your change modifies one or more OpenLineage [facets](https://github.com/OpenLineage/OpenLineage/tree/main/spec/facets)

If you're contributing a new integration, please specify the scope of the integration and how/where it has been tested (e.g., Apache Spark integration supports `S3` and `GCS` filesystem operations, tested with AWS EMR).

### Checklist

- [X] You've [signed-off](https://github.com/OpenLineage/OpenLineage/blob/main/why-the-dco.md) your work
- [X] Your pull request title follows our [guidelines](https://github.com/OpenLineage/OpenLineage/blob/main/CONTRIBUTING.md#creating-pull-requests)
- [X] Your changes are accompanied by tests (_if relevant_)
- [X] Your change contains a [small diff](https://kurtisnusbaum.medium.com/stacked-diffs-keeping-phabricator-diffs-small-d9964f4dcfa6) and is self-contained
- [X] You've updated any relevant documentation (_if relevant_)
- [X] Your comment includes a one-liner for the changelog about the specific purpose of the change (_not required for changes to tests, docs, or CI config_)
- [X] You've versioned the core OpenLineage model or facets according to [SchemaVer](https://docs.snowplowanalytics.com/docs/pipeline-components-and-applications/iglu/common-architecture/schemaver) (_if relevant_)
- [X] You've added a [header](https://github.com/OpenLineage/OpenLineage/tree/main/.github/header_templates.md) to source files (_if relevant_)

----
SPDX-License-Identifier: Apache-2.0\
Copyright 2018-2024 contributors to the OpenLineage project

### CODE REVIEW

File: spec/facets/ColumnLineageDatasetFacet.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "$schema": "https://json-schema.org/draft/2020-12/schema",
-  "$id": "https://openlineage.io/spec/facets/1-1-0/ColumnLineageDatasetFacet.json",
+  "$id": "https://openlineage.io/spec/facets/1-2-0/ColumnLineageDatasetFacet.json",
```
Comment (pawel-big-lebowski) [inline spec/facets/ColumnLineageDatasetFacet.json:3]: Spec changes should be done in separate PRs

Comment (arturowczarek) [reply spec/facets/ColumnLineageDatasetFacet.json:3]: https://github.com/OpenLineage/OpenLineage/pull/3098

File: integration/spark/shared/src/test/java/io/openlineage/spark/agent/lifecycle/plan/column/ColumnLevelLineageBuilderDeprecatedMechanismTest.java
Comment (tnazarew) [inline integration/spark/shared/src/test/java/io/openlineage/spark/agent/lifecycle/plan/column/ColumnLevelLineageBuilderDeprecatedMechanismTest.java:1]: same here

File: integration/spark/spark3/src/main/java/io/openlineage/spark3/agent/lifecycle/plan/column/ColumnLevelLineageUtils.java [hunk 2]
```diff
@@ -55,7 +55,21 @@ public static Optional<OpenLineage.ColumnLineageDatasetFacet> buildColumnLineage
...
-    facetBuilder.fields(context.getBuilder().build());
+    boolean datasetLineageEnabled =
+        context
+            .getOlContext()
+            .getOpenLineageConfig()
+            .getColumnLineageConfig()
+            .isDatasetLineageEnabled();
+    if (!datasetLineageEnabled) {
+      log.warn(
+          "DEPRECATION WARNING: The columnLineage.datasetLineageEnabled configuration is set to false. This flag will default to false in the future versions and soon will be removed (defaulting to true).");
```
Comment (tnazarew) [inline integration/spark/spark3/src/main/java/io/openlineage/spark3/agent/lifecycle/plan/column/ColumnLevelLineageUtils.java:66]: good deprecation message, should it be `warn` though? It's just information so maybe `info`, `WARNING` in the logs should suggest that something could fail here

**Review by tnazarew — CHANGES_REQUESTED** [review]

File: integration/spark/spark3/src/main/java/io/openlineage/spark3/agent/lifecycle/plan/column/ColumnLevelLineageUtils.java (hunk 2 above)
Comment (arturowczarek) [reply integration/spark/spark3/src/main/java/io/openlineage/spark3/agent/lifecycle/plan/column/ColumnLevelLineageUtils.java:66]: It is going to fail in just a couple of releases.

Comment (tnazarew) [reply integration/spark/spark3/src/main/java/io/openlineage/spark3/agent/lifecycle/plan/column/ColumnLevelLineageUtils.java:66]: so in couple of releases it will be a good warning :)

Comment (arturowczarek) [reply integration/spark/spark3/src/main/java/io/openlineage/spark3/agent/lifecycle/plan/column/ColumnLevelLineageUtils.java:66]: Now it warns that the default is going to change and your events will be affected

Comment (tnazarew) [reply integration/spark/spark3/src/main/java/io/openlineage/spark3/agent/lifecycle/plan/column/ColumnLevelLineageUtils.java:66]: I mean, it's not that big of a deal since it will be removed relatively soon. 

Nonetheless warning should tell you about something that may cause failure, so I wouldn't log warnings about something concerning future versions (which current users may not upgrade to anytime soon)

**Review by tnazarew — APPROVED** [review]

File: website/docs/integrations/spark/configuration/spark_conf.md [hunk 3]
```diff
@@ -6,20 +6,21 @@ title: Spark Config Parameters
...
+| spark.openlineage.job.owners.<ownership-type\>        | Specifies ownership of the job. Multiple entries with different types are allowed. Config key name and value are used to create job ownership type and name (available since 1.13).                                                                                                                                                                                                                 | spark.openlineage.job.owners.team="Some Team" |
+| spark.openlineage.columnLineage.datasetLineageEnabled | **Since 1.24.0** Makes the dataset dependencies to be included in their own property `dataset` in the column lineage pattern. If this flag is set to `false`, then the dataset dependencies are merged into `fields` property. The default value is `false` but soon the default will change to `true` and soon after this configuration will be removed. **It is recommended to set it to `true`** | true                                          |
```
Comment (pawel-big-lebowski) [inline website/docs/integrations/spark/configuration/spark_conf.md:26]: There is no need to mention the docs this way anymore.

File: integration/spark/app/src/test/java/io/openlineage/spark/agent/column/ColumnLevelLineageIcebergTest.java [hunk 4]
```diff
@@ -378,7 +380,9 @@ void testCTEQuery() {
     assertColumnDependsOn(facet, "c", FILE, T1_EXPECTED_NAME, "a");
     assertColumnDependsOn(facet, "d", FILE, T1_EXPECTED_NAME, "b");
     assertColumnDependsOnInputs(facet, "c", 1);
-    assertColumnDependsOnInputs(facet, "d", 2);
+    assertColumnDependsOnInputs(facet, "d", 1);
+    assertDatasetDependsOnType(
```
Comment (pawel-big-lebowski) [inline integration/spark/app/src/test/java/io/openlineage/spark/agent/column/ColumnLevelLineageIcebergTest.java:384]: Is this the only test affected in the class by dataset depenendency feature? If so, why do we want to create whole `ColumnLevelLineageIcebergOnlyFieldDependenciesTest` file?

File: integration/spark/shared/src/main/java/io/openlineage/spark/api/SparkOpenLineageConfig.java [hunk 5]
```diff
@@ -32,15 +34,18 @@ public class SparkOpenLineageConfig extends OpenLineageConfig<SparkOpenLineageCo
...
+  private String parentJobNamespace;
+  private String parentRunId;
+  private String overriddenAppName;
+  @NonNull private String debugFacet;
+  private String testExtensionProvider;
+  private JobNameConfig jobName;
+  private JobConfig job;
+
+  @JsonProperty("columnLineage")
+  private ColumnLineageConfig columnLineageConfig;
```
Comment (pawel-big-lebowski) [inline integration/spark/shared/src/main/java/io/openlineage/spark/api/SparkOpenLineageConfig.java:48]: Would it make sense to have it as a part of `DatasetConfig`?

**Review by pawel-big-lebowski — COMMENTED** [review]
I think the original idea behind PR was: let's copy existing tests, bcz there will be a lot of tests changes when implementing dataset lineage feature. The problem with this reasoning is that it turned out only 4 existing tests are affected by the new feature, but going this way requires copying 900 lines of code. 

Is my understanding correct?

File: integration/spark/shared/src/main/java/io/openlineage/spark/api/SparkOpenLineageConfig.java (hunk 5 above)
Comment (arturowczarek) [reply integration/spark/shared/src/main/java/io/openlineage/spark/api/SparkOpenLineageConfig.java:48]: It is often said that nothing is more permanent than s temporary change, but I hope it won't be this case. Moving it to `DatasetConfig` would require extending this class for Spark integration (because we don't have this facet in Java) and it carries the risk introducing some regression. Additionally, it would take extra 30-60 minutes of extra work that could be better spent elsewhere. Is it worth it?

File: integration/spark/app/src/test/java/io/openlineage/spark/agent/column/ColumnLevelLineageIcebergTest.java (hunk 4 above)
Comment (arturowczarek) [reply integration/spark/app/src/test/java/io/openlineage/spark/agent/column/ColumnLevelLineageIcebergTest.java:384]: I chose to copy the files for two reasons. First, to finally remove it, we just delete the entire files without thinking. Second, and most importantly, there is only one place in the tests where the new dataset dependency mode flag is set. Switching it would affect every test. Without hard-to-review to review refactor, it would be impossible to control the flag's value. Additionally, at some point, we'd need to reverse the refactor to avoid causing confusion. The `OnlyFieldDependenciesTest` suites are just copies of what was there so far and are intended for removal.

**Review by pawel-big-lebowski — APPROVED** [review]
